### PR TITLE
(GH-307) Fix path resolution on mac and *nix

### PR DESCRIPTION
**Pull request by jpogran**

### CODE REVIEW

**Comment by glennsarti** [comment]
So I tested this on my Ubuntu desktop and it worked.  Although it wasn't broken before.  Did we expect all *nix users to be broken?

**Comment by jpogran** [comment]
I updated the issue with a little more information

File: src/configuration.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ export class ConnectionConfiguration implements IConnectionConfiguration {
     this.debugFilePath = this.config['languageserver']['debugFilePath'];
   }
 
-  get puppetAgentDir(): string {
+  get puppetBaseDir(): string {
```
Comment (jpogran) [inline src/configuration.ts]: I didn't have to change the name of the accessor here, but it helped keep things straight in my head.

File: src/configuration.ts [hunk 2]
```diff
@@ -37,9 +37,12 @@ export class ConnectionConfiguration implements IConnectionConfiguration {
         if (process.env['PROCESSOR_ARCHITEW6432'] === 'AMD64') {
           programFiles = process.env['ProgramW6432'] || 'C:\\Program Files';
         }
+        // On Windows we have a subfolder called 'Puppet' that has
+        // every product underneath
         return path.join(programFiles, 'Puppet Labs', 'Puppet');
       default:
-        return '/opt/puppetlabs/puppet';
+        // On *nix we don't have a sub folder called 'Puppet'
+        return '/opt/puppetlabs';
```
Comment (jpogran) [inline src/configuration.ts]: This was `/opt/puppetlabs/puppet` before, which would make `puppetDir` `/opt/puppetlabs/puppet/puppet` which doesn't exist

File: src/configuration.ts [hunk 3]
```diff
@@ -61,34 +64,45 @@ export class ConnectionConfiguration implements IConnectionConfiguration {
...
+    return path.join(this.puppetBaseDir, 'mcollective');
   }
 
   get rubydir(): string {
-    return path.join(this.puppetAgentDir, 'sys', 'ruby');
+    switch(process.platform){
+      case 'win32':
+        return path.join(this.puppetBaseDir, 'sys', 'ruby');
+      default:
+        return path.join(this.puppetBaseDir, 'lib', 'ruby');
```
Comment (jpogran) [inline src/configuration.ts]: The vendored ruby paths are different on Mac and *nix, so use the correct path here

File: src/configuration.ts [hunk 4]
```diff
@@ -61,34 +64,45 @@ export class ConnectionConfiguration implements IConnectionConfiguration {
...
-      path.join(this.facterDir, 'lib') +
-      this.pathEnvSeparator() +
-      path.join(this.hieraDir, 'lib') +
-      this.pathEnvSeparator() +
-      path.join(this.mcoDir, 'lib');
+    var p = new Array(
+      path.join(this.puppetDir, 'lib'),
+      path.join(this.facterDir, 'lib'),
+      // path.join(this.hieraDir, 'lib'),
+    ).join(this.pathEnvSeparator());
```
Comment (jpogran) [inline src/configuration.ts]: Reduced the amount of lines needed for this by using an Array.join

File: src/configuration.ts [hunk 5]
```diff
@@ -98,31 +112,16 @@ export class ConnectionConfiguration implements IConnectionConfiguration {
...
-      this.pathEnvSeparator()
-    );
+    return new Array(
+      path.join(this.puppetDir, 'bin'),
+      path.join(this.facterDir, 'bin'),
+      // path.join(this.hieraDir, 'bin'),
+      path.join(this.puppetBaseDir, 'bin'),
+      path.join(this.rubydir, 'bin'),
+      path.join(this.puppetBaseDir, 'sys', 'tools', 'bin')
+    ).join(this.pathEnvSeparator());
```
Comment (jpogran) [inline src/configuration.ts]: Reduced the amount of lines needed for this by using an Array.join

File: src/rubyHelper.ts [hunk 6]
```diff
@@ -52,7 +52,7 @@ export class RubyHelper {
       result.options.env.RUBYLIB = '';
     }
     result.options.env.RUBY_DIR = connectionConfiguration.rubydir;
-    result.options.env.PATH =connectionConfiguration.environmentPath + result.options.env.PATH;
+    result.options.env.PATH =connectionConfiguration.environmentPath + this.pathEnvSeparator() + result.options.env.PATH;
```
Comment (jpogran) [inline src/rubyHelper.ts]: The missing path separator caused problems

**Comment by glennsarti** [comment]
Seems Arch is supported by VSCode...https://aur.archlinux.org/packages/visual-studio-code-bin/?comments=all

File: src/rubyHelper.ts (hunk 6 above)
Comment (glennsarti) [reply src/rubyHelper.ts]: nit: should there be a space between `=connectionC`

**Comment by glennsarti** [comment]
It's odd.  I can't replicate the bad paths with my Ubuntu Desktop.  Another user I had try with Mac is not experiencing the bad pathing.  And I can't see anything in the Puppet repos/docs to suggest a path change.  This PR looks fine and I'm almost about to press the merge button, but I'm having trouble repro-ing.

`no issues from what I can see. VSC 1.25.0, PE 4.10.9. im on OSX 10.12.6, so not high sierra`

**Pull request merged**